### PR TITLE
builtin cd: print error about broken symlink

### PR DESCRIPTION
**Pull request by krobelus**
## Description

Fixes issue #8264

## TODOs:
- [x] Tests have been added for regressions fixed
- [x] User-visible changes noted in CHANGELOG.rst

### CODE REVIEW

File: src/wutil.cpp [hunk 1]
```diff
@@ -203,6 +203,27 @@ int make_fd_blocking(int fd) {
...
 }
 
+maybe_t<wcstring> wreadlink(const wcstring &file_name) {
+    struct stat buf;
+    if (lwstat(file_name, &buf) == -1) {
+        return none();
+    }
+    ssize_t bufsize = buf.st_size + 1;
+    char target_buf[bufsize];
+    const std::string tmp = wcs2string(file_name);
```
Comment (faho) [inline src/wutil.cpp:213]: (note: We really really do way too much of this wcs2string and back - this will do wcs2string for the `lwstat` above, and then wcs2string again for the readlink. In some benchmarks ~50% of the time is taken in widening and narrowing strings)

(but since this is just for the error, it's okay for now - we should just probably start doing all the path stuff in narrow exclusively, so you can keep one narrow string around if you need to do more than one thing)

File: src/wutil.cpp [hunk 2]
```diff
@@ -705,9 +726,7 @@ file_id_t file_id_for_fd(int fd) {
     return result;
 }
 
-file_id_t file_id_for_fd(const autoclose_fd_t &fd) {
-    return file_id_for_fd(fd.fd());
-}
+file_id_t file_id_for_fd(const autoclose_fd_t &fd) { return file_id_for_fd(fd.fd()); }
```
Comment (faho) [inline src/wutil.cpp:729]: This line wasn't touched, at all. That's the issue with running clang-format on save - you touch a lot of unrelated code.

We should probably do another formatting run (but changes like this are the worst - clang-format is super anal about having lines exactly the correct way, which is just useless churn and should be disabled, but I've so far not found a way)

Comment (krobelus) [reply src/wutil.cpp:729]: > but changes like this are the worst - clang-format is super anal about having lines exactly the correct way, which is just useless churn and should be disabled, but I've so far not found a way

I'm not sure what you mean exactly - it's the formatter's job to be precise, so we don't have to.
This style can be configured with `AllowShortFunctionsOnASingleLine`, see also https://zed0.co.uk/clang-format-configurator/. I somehow have lots of energy for evangelizing clang-format.. it makes refactorings much easier and avoids thinking or communicating about style.

Assuming we want this change, I'd probably still do it because it's small and I'm already touching the file / invalidating builds. Though that's really not a strong reason.

Comment (faho) [reply src/wutil.cpp:729]: The problem is that it's *annoyingly* precise.

We've had cases where it flip-flopped. E.g. rename this variable to "file_desc" and it might just decide that it *has to* be split. That's reformatting that basically helps nobody. Line length really only matters if lines constantly go over by a lot which requires you to scroll to view the lines you care about.

And because this particular kind of reformatting isn't whitespace-only, `git blame -w` won't ignore it and will show you the reformatting commit first, so you then need to run it again with `somehash^`. There's a way to give `git blame` commits to ignore, but that also only really works if the commits only reformat (or you could be ignoring a functional change!).

So I'm really quite annoyed by clang-format's super precise insistence on line length and would much prefer if it had some sort of slack where it keeps whatever's there, because as it stands it either forces joined style or forces separated style. (that's what I meant by "disabling" it - currently you can only say "always separate" or "separate exactly when clang-format decides to")

And if we can't have that, we could at least restyle the entire codebase every so often in separate commits and add their hashes to some file so I can run `git blame -w --ignore-revs-file :/restylecommits.txt` (or set "blame.ignoreRevsFile") to ignore them, and stop restyling unrelated stuff in regular commits.

Comment (faho) [reply src/wutil.cpp:729]: Note: I don't particularly care about this one occurence, it's more that I don't want commits to have unrelated formatting changes so `git blame` is nicer to use, and that I consider clang-format to not be a great autoformatter because it is annoyingly precise, which causes it to add a bunch of formatting churn.
